File: example/cpp/preview_depth.cpp
/**
 * @file preview_depth.cpp
 * @brief Arducam TOF Depth Camera Preview Application
 * 
 * This application demonstrates how to capture and display depth data from the Arducam TOF camera.
 * It provides real-time depth visualization with confidence filtering, interactive region selection,
 * and the ability to save depth data to files.
 * 
 * Features:
 * - Real-time depth frame capture and display
 * - Confidence-based filtering for noise reduction
 * - Depth threshold filtering to remove far objects
 * - Interactive mouse selection for distance measurement
 * - FPS monitoring and display
 * - Depth data saving functionality
 * - Color-mapped depth visualization
 * - Adjustable depth threshold with trackbar control
 * 
 * @author Arducam
 * @version 1.0
 * @date 2024
 */

#include "ArducamTOFCamera.hpp"
#include <chrono>
#include <fstream>
#include <iostream>
#include <opencv2/core.hpp>
#include <opencv2/highgui.hpp>
#include <opencv2/imgproc.hpp>

using namespace Arducam;

// ============================================================================
// Configuration Constants
// ============================================================================

/**
 * @brief Maximum distance range for the TOF camera in millimeters
 * 
 * Valid values are 2000 (2 meters) or 4000 (4 meters).
 * This affects the camera's measurement range and accuracy.
 */
#define MAX_DISTANCE 4000

// ============================================================================
// Global Variables
// ============================================================================

/**
 * @brief Rectangle for selected region (where user clicked)
 * Used for distance measurement at specific points
 */
cv::Rect seletRect(0, 0, 0, 0);

/**
 * @brief Rectangle for mouse follow region (current mouse position)
 * Provides visual feedback of current mouse position
 */
cv::Rect followRect(0, 0, 0, 0);

/**
 * @brief Maximum width of the camera frame
 * Updated dynamically based on camera resolution
 */
int max_width = 240;

/**
 * @brief Maximum height of the camera frame
 * Updated dynamically based on camera resolution
 */
int max_height = 180;

/**
 * @brief Current maximum range setting of the camera
 * Retrieved from camera control settings
 */
int max_range = 0;

/**
 * @brief Confidence threshold for depth filtering
 * Values below this threshold are considered unreliable and filtered out
 */
int confidence_value = 30;

/**
 * @brief Depth threshold for filtering far objects
 * Values above this threshold (in millimeters) are filtered out and displayed as black
 */
int depth_threshold = 2000;

// ============================================================================
// Callback Functions
// ============================================================================

/**
 * @brief Callback function for confidence threshold changes
 * 
 * This function is called when the confidence trackbar value changes.
 * Currently not implemented but can be used to dynamically adjust
 * the confidence filtering threshold.
 * 
 * @param pos New confidence threshold value (0-255)
 * @param userdata User data pointer (unused)
 */
void on_confidence_changed(int pos, void* userdata)
{
    // TODO: Implement dynamic confidence threshold adjustment
    // confidence_value = pos;
}

/**
 * @brief Callback function for depth threshold changes
 * 
 * This function is called when the depth threshold trackbar value changes.
 * It updates the global depth threshold for filtering far objects.
 * 
 * @param pos New depth threshold value (0-4000mm)
 * @param userdata User data pointer (unused)
 */
void on_depth_threshold_changed(int pos, void* userdata)
{
    depth_threshold = pos;
}

// ============================================================================
// Utility Functions
// ============================================================================

/**
 * @brief Display and calculate FPS (Frames Per Second)
 * 
 * This function tracks the number of frames processed and calculates
 * the current FPS, displaying it every second. Uses high-resolution
 * timing for accurate measurements.
 * 
 * The FPS is calculated by counting frames over a 1-second interval
 * and resetting the counter after each display.
 */
void display_fps(void)
{
    using std::chrono::high_resolution_clock;
    using namespace std::literals;
    static int count = 0;
    static auto time_beg = high_resolution_clock::now();
    auto time_end = high_resolution_clock::now();
    ++count;
    auto duration_ms = (time_end - time_beg) / 1ms;
    if (duration_ms >= 1000) {
        std::cout << "fps:" << count << std::endl;
        count = 0;
        time_beg = time_end;
    }
}

/**
 * @brief Save depth image data to a binary file
 * 
 * Saves the current depth frame as a raw binary file with a timestamped filename.
 * The file format is: depth_[width]_[height]_f32_[timestamp].raw
 * 
 * @param image Pointer to the depth image data (float array)
 * @param width Width of the depth image
 * @param height Height of the depth image
 */
void save_image(float* image, int width, int height)
{
    using namespace std::literals;
    // filename = "depth_$width$_$height$_f32_$time.raw"
    auto now = std::chrono::system_clock::now().time_since_epoch() / 1ms;
    std::string filename =
        "depth_" + std::to_string(width) + "_" + std::to_string(height) + "_f32_" + std::to_string(now) + ".raw";
    std::ofstream file(filename, std::ios::binary);
    file.write(reinterpret_cast<char*>(image), width * height * sizeof(float));
    file.close();
}

/**
 * @brief Rotate an OpenCV Mat 180 degrees clockwise
 * 
 * Performs a 180-degree rotation by flipping the image both horizontally and vertically.
 * This is useful for correcting camera orientation if needed.
 * 
 * @param src Source image matrix to rotate
 * @return Rotated image matrix
 */
cv::Mat matRotateClockWise180(cv::Mat src)
{
    if (src.empty()) {
        std::cerr << "RorateMat src is empty!";
    }

    flip(src, src, 0);  // Flip vertically
    flip(src, src, 1);  // Flip horizontally
    return src;
}

/**
 * @brief Apply confidence filtering to grayscale preview image
 * 
 * This function filters out pixels with low confidence values by setting them to white (255).
 * It iterates through each pixel and compares the confidence value against the threshold.
 * 
 * @param preview_ptr Grayscale preview image to filter
 * @param amplitude_image_ptr Confidence/amplitude image for filtering
 */
void getPreview(cv::Mat preview_ptr, cv::Mat amplitude_image_ptr)
{
    auto len = preview_ptr.rows * preview_ptr.cols;
    for (int line = 0; line < preview_ptr.rows; line++) {
        for (int col = 0; col < preview_ptr.cols; col++) {
            if (amplitude_image_ptr.at<float>(line, col) < confidence_value)
                preview_ptr.at<uint8_t>(line, col) = 255;
        }
    }
}

/**
 * @brief Apply confidence filtering to RGB preview image
 * 
 * This function filters out pixels with low confidence values by setting them to black (0,0,0).
 * Uses OpenCV's setTo function for efficient vectorized operations.
 * 
 * @param preview_ptr RGB preview image to filter
 * @param amplitude_image_ptr Confidence/amplitude image for filtering
 */
void getPreviewRGB(cv::Mat preview_ptr, cv::Mat amplitude_image_ptr)
{
    preview_ptr.setTo(cv::Scalar(0, 0, 0), amplitude_image_ptr < confidence_value);
    // Optional: Apply Gaussian blur for noise reduction
    // cv::GaussianBlur(preview_ptr, preview_ptr, cv::Size(7, 7), 0);
}

/**
 * @brief Create depth threshold filtered image
 * 
 * This function creates a filtered depth image where pixels with depth values
 * greater than the threshold are set to black, effectively filtering out far objects.
 * 
 * @param depth_frame Input depth frame (CV_32F)
 * @param confidence_frame Input confidence frame for additional filtering
 * @return Filtered depth image (CV_8U)
 */
cv::Mat createDepthThresholdImage(cv::Mat depth_frame, cv::Mat confidence_frame)
{
    // Create a copy of the depth frame for filtering
    cv::Mat filtered_depth = depth_frame.clone();
    
    // Apply depth threshold filtering - set far pixels to 0
    filtered_depth.setTo(0, depth_frame > depth_threshold);
    
    // Also apply confidence filtering
    filtered_depth.setTo(0, confidence_frame < confidence_value);
    
    // Convert to 8-bit for display
    cv::Mat result;
    filtered_depth.convertTo(result, CV_8U, 255.0 / 7000, 0);
    
    // Apply rainbow color map for better visualization
    cv::applyColorMap(result, result, cv::COLORMAP_RAINBOW);
    
    return result;
}

/**
 * @brief Mouse callback function for interactive region selection
 * 
 * Handles mouse events for selecting regions of interest in the depth image.
 * Creates 8x8 pixel rectangles for distance measurement and visual feedback.
 * 
 * Mouse Events:
 * - Left button down: Start selection (currently unused)
 * - Left button up: Set selected rectangle for distance measurement
 * - Mouse move: Update follow rectangle for visual feedback
 * 
 * @param event Mouse event type (click, move, etc.)
 * @param x Mouse x-coordinate
 * @param y Mouse y-coordinate
 * @param flags Additional mouse flags (unused)
 * @param param User parameter (unused)
 */
void onMouse(int event, int x, int y, int flags, void* param)
{
    // Ensure mouse position is within valid bounds (4 pixels from edges)
    if (x < 4 || x > (max_width - 4) || y < 4 || y > (max_height - 4))
        return;
        
    switch (event) {
    case cv::EVENT_LBUTTONDOWN:
        // Start of selection (currently not used)
        break;

    case cv::EVENT_LBUTTONUP:
        // Set selected rectangle for distance measurement
        seletRect.x = x - 4 ? x - 4 : 0;
        seletRect.y = y - 4 ? y - 4 : 0;
        seletRect.width = 8;
        seletRect.height = 8;
        break;
    default:
        // Update follow rectangle for visual feedback
        followRect.x = x - 4 ? x - 4 : 0;
        followRect.y = y - 4 ? y - 4 : 0;
        followRect.width = 8;
        followRect.height = 8;
        break;
    }
}

// ============================================================================
// Main Application
// ============================================================================

/**
 * @brief Main function for the Arducam TOF Depth Camera Preview Application
 * 
 * This function initializes the TOF camera, sets up the display windows,
 * and runs the main processing loop for real-time depth visualization.
 * 
 * The application flow:
 * 1. Initialize and open the TOF camera
 * 2. Start depth frame capture
 * 3. Configure camera settings (range, etc.)
 * 4. Set up OpenCV windows, trackbars, and mouse callbacks
 * 5. Run main processing loop with real-time filtering
 * 6. Clean up resources on exit
 * 
 * Windows:
 * - "preview": Main depth visualization with mouse interaction
 * - "confidence": Confidence/amplitude data display
 * - "depth_threshold": Depth threshold filtered view with adjustable trackbar
 * 
 * @return 0 on success, -1 on error
 */
int main()
{
    // ========================================================================
    // Camera Initialization
    // ========================================================================
    
    ArducamTOFCamera tof;
    ArducamFrameBuffer* frame;
    
    // Open camera connection (CSI interface)
    if (tof.open(Connection::CSI, 0)) {
        std::cerr << "Failed to open camera" << std::endl;
        return -1;
    }

    // Start depth frame capture
    if (tof.start(FrameType::DEPTH_FRAME)) {
        std::cerr << "Failed to start camera" << std::endl;
        return -1;
    }
    
    // Configure camera range setting
    // Note: Modify MAX_DISTANCE constant to change range
    tof.setControl(Control::RANGE, MAX_DISTANCE);
    tof.getControl(Control::RANGE, &max_range);
    
    // Get camera information
    auto info = tof.getCameraInfo();
    std::cout << "open camera with (" << info.width << "x" << info.height << ")" << std::endl;

    // ========================================================================
    // Display Setup
    // ========================================================================
    
    // Allocate buffer for preview image
    uint8_t* preview_ptr = new uint8_t[info.width * info.height * 2];
    
    // Create OpenCV windows
    cv::namedWindow("preview", cv::WINDOW_AUTOSIZE);
    cv::setMouseCallback("preview", onMouse);
    
    // Create depth threshold filtering window
    cv::namedWindow("depth_threshold", cv::WINDOW_AUTOSIZE);
    
    // Create trackbar for depth threshold (0-4000mm)
    cv::createTrackbar("Depth Threshold (mm)", "depth_threshold", &depth_threshold, 4000, on_depth_threshold_changed);

    // ========================================================================
    // Main Processing Loop
    // ========================================================================
    
    for (;;) {
        // Request a new frame from the camera (200ms timeout)
        Arducam::FrameFormat format;
        frame = tof.requestFrame(200);
        if (frame == nullptr) {
            continue;  // Skip this iteration if no frame available
        }
        
        // Get frame format information
        frame->getFormat(FrameType::DEPTH_FRAME, format);
        std::cout << "frame: (" << format.width << "x" << format.height << ")" << std::endl;
        max_height = format.height;
        max_width = format.width;

        // Extract depth and confidence data from frame
        float* depth_ptr = (float*)frame->getData(FrameType::DEPTH_FRAME);
        float* confidence_ptr = (float*)frame->getData(FrameType::CONFIDENCE_FRAME);

        // Create OpenCV matrices from raw data
        cv::Mat result_frame(format.height, format.width, CV_8U, preview_ptr);
        cv::Mat depth_frame(format.height, format.width, CV_32F, depth_ptr);
        cv::Mat confidence_frame(format.height, format.width, CV_32F, confidence_ptr);

        // Optional: Apply 180-degree rotation if needed
        // depth_frame = matRotateClockWise180(depth_frame);
        // result_frame = matRotateClockWise180(result_frame);
        // confidence_frame = matRotateClockWise180(confidence_frame);
        
        // Convert depth data to 8-bit for visualization (scale to 7000mm max)
        depth_frame.convertTo(result_frame, CV_8U, 255.0 / 7000, 0);

        // Apply rainbow color map for better depth visualization
        cv::applyColorMap(result_frame, result_frame, cv::COLORMAP_RAINBOW);
        
        // Apply confidence filtering to remove unreliable pixels
        getPreviewRGB(result_frame, confidence_frame);

        // Convert confidence data to 8-bit for display
        confidence_frame.convertTo(confidence_frame, CV_8U, 255.0 / 1024, 0);

        // Display confidence window
        cv::imshow("confidence", confidence_frame);

        // Create and display depth threshold filtered image
        cv::Mat depth_threshold_image = createDepthThresholdImage(depth_frame, confidence_frame);
        cv::imshow("depth_threshold", depth_threshold_image);

        // Draw selection rectangles for distance measurement
        cv::rectangle(result_frame, seletRect, cv::Scalar(0, 0, 0), 2);      // Black rectangle for selected area
        cv::rectangle(result_frame, followRect, cv::Scalar(255, 255, 255), 1); // White rectangle for mouse follow

        // Calculate and display average distance in selected region
        std::cout << "select Rect distance: " << cv::mean(depth_frame(seletRect)).val[0] << std::endl;

        // Display main preview window
        cv::imshow("preview", result_frame);

        // Handle keyboard input
        auto key = cv::waitKey(1);
        if (key == 27 || key == 'q') {
            break;  // Exit on ESC or 'q'
        } else if (key == 's') {
            save_image(depth_ptr, format.width, format.height);  // Save depth data
        }
        
        // Update FPS counter
        display_fps();
        
        // Release frame buffer
        tof.releaseFrame(frame);
    }

    // ========================================================================
    // Cleanup and Exit
    // ========================================================================
    
    // Stop camera capture
    if (tof.stop()) {
        return -1;
    }

    // Close camera connection
    if (tof.close()) {
        return -1;
    }

    // Clean up allocated memory
    delete[] preview_ptr;

    return 0;
}

File: example/python/preview_depth.py
"""
Arducam TOF Depth Camera Preview Application

This application demonstrates how to capture and display depth data from the Arducam TOF camera
using Python. It provides real-time depth visualization with confidence filtering, interactive
region selection, and distance measurement capabilities.

Features:
- Real-time depth frame capture and display
- Confidence-based filtering for noise reduction
- Interactive mouse selection for distance measurement
- Color-mapped depth visualization
- Support for both VGA and HQVGA camera types
- Dynamic confidence threshold adjustment

Author: Arducam
Version: 1.0
Date: 2024
"""

import cv2
import numpy as np
import ArducamDepthCamera as ac

# ============================================================================
# Configuration Constants
# ============================================================================

# MAX_DISTANCE value modifiable is 2000 or 4000
# 2000 = 2 meters range, 4000 = 4 meters range
MAX_DISTANCE = 4000


# ============================================================================
# Utility Classes
# ============================================================================

class UserRect:
    """
    Utility class for managing rectangular regions selected by the user.
    
    This class provides convenient properties for accessing rectangle coordinates
    in different formats (OpenCV rectangle format, numpy slice format) and
    checking if the rectangle is empty.
    """
    
    def __init__(self) -> None:
        """Initialize an empty rectangle."""
        self.start_x = 0
        self.start_y = 0
        self.end_x = 0
        self.end_y = 0

    @property
    def rect(self):
        """
        Get rectangle in OpenCV format (x, y, width, height).
        
        Returns:
            tuple: (x, y, width, height) for use with cv2.rectangle()
        """
        return (
            self.start_x,
            self.start_y,
            self.end_x - self.start_x,
            self.end_y - self.start_y,
        )

    @property
    def slice(self):
        """
        Get rectangle as numpy slice for array indexing.
        
        Returns:
            tuple: (row_slice, col_slice) for use with numpy arrays
        """
        return (slice(self.start_y, self.end_y), slice(self.start_x, self.end_x))

    @property
    def empty(self):
        """
        Check if the rectangle is empty (has zero area).
        
        Returns:
            bool: True if rectangle is empty, False otherwise
        """
        return self.start_x == self.end_x and self.start_y == self.end_y


# ============================================================================
# Global Variables
# ============================================================================

# Confidence threshold for depth filtering
# Values below this threshold are considered unreliable and filtered out
confidence_value = 30

# Rectangle objects for user interaction
selectRect, followRect = UserRect(), UserRect()

# ============================================================================
# Utility Functions
# ============================================================================

def getPreviewRGB(preview: np.ndarray, confidence: np.ndarray) -> np.ndarray:
    """
    Apply confidence filtering to RGB preview image.
    
    This function filters out pixels with low confidence values by setting them to black.
    It also handles NaN values that might be present in the depth data.
    
    Args:
        preview (np.ndarray): RGB preview image to filter
        confidence (np.ndarray): Confidence/amplitude image for filtering
        
    Returns:
        np.ndarray: Filtered RGB preview image
    """
    preview = np.nan_to_num(preview)  # Handle NaN values
    preview[confidence < confidence_value] = (0, 0, 0)  # Set low confidence pixels to black
    return preview


def on_mouse(event, x, y, flags, param):
    """
    Mouse callback function for interactive region selection.
    
    Handles mouse events for selecting regions of interest in the depth image.
    Creates 8x8 pixel rectangles for distance measurement and visual feedback.
    
    Args:
        event: Mouse event type (click, move, etc.)
        x: Mouse x-coordinate
        y: Mouse y-coordinate
        flags: Additional mouse flags (unused)
        param: User parameter (unused)
    """
    global selectRect, followRect

    if event == cv2.EVENT_LBUTTONDOWN:
        # Start of selection (currently not used)
        pass

    elif event == cv2.EVENT_LBUTTONUP:
        # Set selected rectangle for distance measurement (8x8 pixels around click point)
        selectRect.start_x = x - 4
        selectRect.start_y = y - 4
        selectRect.end_x = x + 4
        selectRect.end_y = y + 4
    else:
        # Update follow rectangle for visual feedback (8x8 pixels around mouse position)
        followRect.start_x = x - 4
        followRect.start_y = y - 4
        followRect.end_x = x + 4
        followRect.end_y = y + 4


def on_confidence_changed(value):
    """
    Callback function for confidence threshold changes.
    
    This function is called when the confidence trackbar value changes.
    It updates the global confidence threshold for depth filtering.
    
    Args:
        value (int): New confidence threshold value (0-255)
    """
    global confidence_value
    confidence_value = value


def usage(argv0):
    """
    Display usage information for the application.
    
    Args:
        argv0 (str): Name of the script (argv[0])
    """
    print("Usage: python " + argv0 + " [options]")
    print("Available options are:")
    print(" -d        Choose the video to use")


# ============================================================================
# Main Application
# ============================================================================

def main():
    """
    Main function for the Arducam TOF Depth Camera Preview Application.
    
    This function initializes the TOF camera, sets up the display windows,
    and runs the main processing loop for real-time depth visualization.
    
    The application flow:
    1. Initialize and open the TOF camera
    2. Start depth frame capture
    3. Configure camera settings (range, etc.)
    4. Set up OpenCV windows and mouse callbacks
    5. Run main processing loop
    6. Clean up resources on exit
    """
    print("Arducam Depth Camera Demo.")
    print("  SDK version:", ac.__version__)

    # ========================================================================
    # Camera Initialization
    # ========================================================================
    
    cam = ac.ArducamCamera()
    cfg_path = None
    # cfg_path = "file.cfg"  # Uncomment to use configuration file

    # Color definitions for rectangle drawing
    black_color = (0, 0, 0)
    white_color = (255, 255, 255)

    # Open camera connection
    ret = 0
    if cfg_path is not None:
        ret = cam.openWithFile(cfg_path, 0)
    else:
        ret = cam.open(ac.Connection.CSI, 0)
    if ret != 0:
        print("Failed to open camera. Error code:", ret)
        return

    # Start depth frame capture
    ret = cam.start(ac.FrameType.DEPTH)
    if ret != 0:
        print("Failed to start camera. Error code:", ret)
        cam.close()
        return

    # Configure camera range setting
    cam.setControl(ac.Control.RANGE, MAX_DISTANCE)
    r = cam.getControl(ac.Control.RANGE)

    # Get camera information
    info = cam.getCameraInfo()
    print(f"Camera resolution: {info.width}x{info.height}")

    # ========================================================================
    # Display Setup
    # ========================================================================
    
    # Create OpenCV windows
    cv2.namedWindow("preview", cv2.WINDOW_AUTOSIZE)
    cv2.setMouseCallback("preview", on_mouse)

    # Add confidence trackbar for VGA cameras (only VGA supports confidence)
    if info.device_type == ac.DeviceType.VGA:
        cv2.createTrackbar(
            "confidence", "preview", confidence_value, 255, on_confidence_changed
        )

    # ========================================================================
    # Main Processing Loop
    # ========================================================================
    
    while True:
        # Request a new frame from the camera (2000ms timeout)
        frame = cam.requestFrame(2000)
        if frame is not None and isinstance(frame, ac.DepthData):
            # Extract depth and confidence data from frame
            depth_buf = frame.depth_data
            confidence_buf = frame.confidence_data

            # Convert depth data to 8-bit for visualization
            result_image = (depth_buf * (255.0 / r)).astype(np.uint8)
            
            # Apply rainbow color map for better depth visualization
            result_image = cv2.applyColorMap(result_image, cv2.COLORMAP_RAINBOW)
            
            # Apply confidence filtering to remove unreliable pixels
            result_image = getPreviewRGB(result_image, confidence_buf)

            # Normalize confidence data for display
            cv2.normalize(confidence_buf, confidence_buf, 1, 0, cv2.NORM_MINMAX)

            # Display confidence window
            cv2.imshow("preview_confidence", confidence_buf)

            # Draw selection rectangles for distance measurement
            cv2.rectangle(result_image, followRect.rect, white_color, 1)  # White rectangle for mouse follow
            if not selectRect.empty:
                cv2.rectangle(result_image, selectRect.rect, black_color, 2)  # Black rectangle for selected area
                print("select Rect distance:", np.mean(depth_buf[selectRect.slice]))

            # Display main preview window
            cv2.imshow("preview", result_image)
            cam.releaseFrame(frame)

        # Handle keyboard input
        key = cv2.waitKey(1)
        if key == ord("q"):
            break  # Exit on 'q' key

    # ========================================================================
    # Cleanup and Exit
    # ========================================================================
    
    # Stop camera capture and close connection
    cam.stop()
    cam.close()


if __name__ == "__main__":
    main()
